File: 2022/03/3.py
def priority(s):
  i = ord(s)
  return i - (96 if (i > 96) else 38)

def overlaps(sets):
  return [s.intersection(*ss).pop() for [s, *ss] in sets]

backpacks = [list(line.strip()) for line in open("input.txt")]

compartments = [[set(bp[0:len(bp) // 2]), set(bp[len(bp) // 2:])] for bp in backpacks]
print("Part 1:", sum(priority(x) for x in overlaps(compartments)))

groups = [map(set, backpacks[i:i + 3]) for i in range(0, len(backpacks), 3)]
print("Part 2:", sum(priority(x) for x in overlaps(groups)))

File: 2022/03/3.clj
(ns aoc.2022.03.3
  (:require [aoc.common :refer [read-input sum]]
            [clojure.set :as set]
            [clojure.string :as str]))

(defn priority [s]
  (let [i (int (.charAt s 0))]
    (- i (if (> i 96) 96 38))))

(defn sum-priorities [sets]
  (->> sets
       (map #(apply set/intersection %))
       (map first)
       (map priority)
       sum))

(let [backpacks (->> (read-input)
                     (map #(str/split % #"")))]
  (->> backpacks
       (map #(partition (/ (count %) 2) %))
       (map #(map set %))
       sum-priorities
       (println "Part 1:"))

  (->> backpacks
       (map set)
       (partition 3)
       sum-priorities
       (println "Part 2:")))